File: src/common/web/data/entities/project/features/ResourcesMetadataFeature.ts
import { Transform, Type, plainToInstance } from "class-transformer";

import { LayoutObject } from "../../../../ui/components/propertyeditor/ProjectObjectStore";
import { ProjectFeature, type ProjectFeatureID } from "./ProjectFeature";

/**
 * @typedef {ResourcesMetadataKey}
 *
 * Represents the keys of the ResourcesMetadata object.
 */
export type ResourcesMetadataKey = keyof typeof ResourcesMetadata;

/**
 * The resources metadata type.
 *
 *
 */
export class ResourcesMetadata {
    [key: string]: LayoutObject;

    public constructor(metadata: { [key: string]: LayoutObject[] } = {}) {
        if (metadata.value === undefined) return;

        Object.entries(metadata.value).forEach(([key, value]: [ResourcesMetadataKey, LayoutObject]) => {
            if (value !== undefined) this[key as ResourcesMetadataKey] = plainToInstance(LayoutObject, value);
        });
    }
}

/**
 * Data class for the files project feature.
 */
export class ResourcesMetadataFeature extends ProjectFeature {
    public static readonly FeatureID: ProjectFeatureID = "resources_metadata";

    // @ts-ignore
    @Type(() => ResourcesMetadata)
    // @ts-ignore
    @Transform((value) => new ResourcesMetadata(value), { toClassOnly: false })
    public readonly metadata: ResourcesMetadata;

    public constructor(resourcesMetadata: ResourcesMetadata = {}) {
        super();

        this.metadata = resourcesMetadata;
    }

    public get featureID(): ProjectFeatureID {
        return ResourcesMetadataFeature.FeatureID;
    }
}

File: src/common/py/data/entities/project/features/resources_metadata_feature.py
import typing
from dataclasses import dataclass, field

from dataclasses_json import dataclass_json

from ...properties.properties_object import ProjectObject
from .project_feature import ProjectFeature, ProjectFeatureID

ResourcesMetadata = typing.Dict[
    str, typing.List[ProjectObject]
    ]



@dataclass_json
@dataclass
class ResourcesMetadataFeature(ProjectFeature):
    """
    Data class for the resources metadata project feature.
    """

    feature_id: typing.ClassVar[ProjectFeatureID] = "resources_metadata"

    metadata: ResourcesMetadata = field(default_factory=dict)
